File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
// var isValid = function(s) {
 
//         let stack=[]
//         let opening=["(","{","["]
//         let closing={")":"(",
//                      "}":"{",
//                      "]":"["}
//         for(let char of s){
//            if(opening.indexOf(char)>=0) stack.push(char)
//             else{
//                 if(closing[char]&&closing[char]===stack[stack.length-1]) stack.pop()
//                 else return false
//             }
        
        
//     }
//     if(stack.length>0) return false
//     return true
    
// };

var isValid = function(s) {
// use stack to keep track of openning parentheses
// iterate through the string, if it's an openting parenthese, push it to the stack
    // if char is a closing parenthese, compare with the last element in stack
    // -if match with the last open parenthese, pop it off the stack and continue

    let stack=[]
    let openingParentheses="([{"
    let closingParentheses={")":"(",
                     "}":"{",
                     "]":"["}
    
    for(let char of s){
        if(openingParentheses.includes(char)) stack.push(char)
        else{
            if(closingParentheses[char] && closingParentheses[char]===stack[stack.length-1]) stack.pop()
            else return false
          
        }
    }
   if(stack.length>0) return false
    return true
    

  
    
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
#         add opening parentheses to stack then pop off when having matching closing parenthese
        # stack=[]
        # openings=["(","[","{"]
        # closings={")":"(","}":"{","]":"["}
        # for char in s:
        #     if char in openings:
        #         stack.append(char)
        #     else:
        #         if stack and closings[char]==stack[-1]:
        #             stack.pop()
        #         else:
        #             return False
        # return True if not stack else False
        
        stack = []
        for char in s:
            # 压栈
            if char == "{" or char == "[" or char == "(":
                stack.append(char)
            else:
                # check if stack is empty
                if not stack: # empty stack, try to push in closing parentheses
                    return False
                # check if the top of stack will match
                if char == "]" and stack[-1] != "[" or char == "}" and stack[-1] != "{" or char == ")" and stack[-1] != "(":
                    return False
                stack.pop()
                # whether stack is empty
        return not stack
        
        
        
        